packages/connect/src/protocol-connect/get-request.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Message, protoBase64 } from "@bufbuild/protobuf";
+import type { AnyMessage } from "@bufbuild/protobuf";
+import type { UnaryRequest } from "../index.js";
+import {
+  headerContentType,
+  headerProtocolVersion,
+  headerUnaryAcceptEncoding,
+  headerUnaryContentLength,
+  headerUnaryEncoding,
+} from "./headers.js";
+
+const contentTypePrefix = "application/";
+
+function encodeMessageForUrl(message: Uint8Array, useBase64: boolean): string {
+  if (useBase64) {
+    // TODO(jchadwick-buf): Three regex replaces seems excessive.
+    // Can we make protoBase64.enc more flexible?
+    return protoBase64
+      .enc(message)
+      .replace(/\+/g, "-")
+      .replace(/\//g, "_")
+      .replace(/=+$/, "");
+  } else {
+    return encodeURIComponent(new TextDecoder().decode(message));
+  }
+}
+
+/**
+ * @private Internal code, does not follow semantic versioning.
+ */
+export function transformConnectPostToGetRequest<
+  I extends Message<I> = AnyMessage,
+  O extends Message<O> = AnyMessage
+>(
+  request: UnaryRequest<I, O>,
+  message: Uint8Array,
+  useBase64: boolean
+): UnaryRequest<I, O> {
+  let query = "?connect=v1";
+  const contentType = request.header.get(headerContentType);
+  if (contentType?.indexOf(contentTypePrefix) === 0) {
+    query +=
+      "&encoding=" +
+      encodeURIComponent(contentType.slice(contentTypePrefix.length));
+  }
+  const compression = request.header.get(headerUnaryEncoding);
+  if (compression !== null && compression !== "identity") {
+    query += "&compression=" + encodeURIComponent(compression);
+
+    // Force base64 for compressed payloads.
+    useBase64 = true;
```

Comment:
I think we are going against the grain with percent encoding arbitrary octets. The standard libs I am familiar with  will expect UTF-8 by default, and data that is not valid UTF-8 will require bespoke code paths, if feasible at all. Case in point are dotnet's [ParseQueryString](https://learn.microsoft.com/en-us/dotnet/api/system.web.httputility.parsequerystring?view=netcore-3.1) and [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) in JS. From my understanding, RFC3986 mandates the use of UTF-8 specifically to avoid character encoding issues between different systems.

Percent-encoding JSON is fine because it is always UTF-8 (RFC8529 section 8.1), but proto binary and compressed messages should IMO always be base64 encoded.

John and @akshayjshah, what do you think?

Reply:
Agreed, and we're working to clarify the spec as discussed offline. The client doesn't try to encode binary as percent-encoding, and we have another PR to discuss this issue in at the moment, so I'll mark this resolved now.